### DISCUSSION TITLE
S3 sensors

### DISCUSSION THREAD

**Discussion by yuval-airis** [dagster-io/dagster#18216] [Q&A]
**Question 1 - S3 op/job trigger**
I wanted to trigger certain ops on new files in S3, and saw that you actually had an example like that in your documentation:

https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#s3-sensors

I understood the logic within the sensor (It checks whether new keys exist in a given S3 bucket returning SkipReason if there aren't and a sequence of RunRequests otherwise), but it isn't clear to me when that sensor is evaluated. (In other words, how long after a file appears in my S3 bucket, would I get triggered?)

**Question 2 - Run Requests**
In addition, I wasn't sure about the output either. I return run requests, each with a run key and run config. The example didn't have the job that was triggered, so it wasn't clear to me what the run requests should contain. (I saw a run key and a run config, but I expected to see inputs for job runs. Are these those?)

Question 3 - 
I looked at the code of [get_s3_keys](https://github.com/dagster-io/dagster/blob/a9336275cf23ff63c0f70c1eaff5034f5bf67497/python_modules/libraries/dagster-aws/dagster_aws/s3/sensor.py#L7) and noticed it reads all keys (cursor is initially "") and only later starts updating it after every partial read. Couldn't cursor have been initialized as since_key? (Line 15)

**Comment by marcilj** [comment]
I'm not part of Dagster Team or anything but here's my answers.

For question 1

>  By default, the Dagster daemon runs a sensor 30 seconds after that sensor's previous evaluation finishes executing. You can configure the interval using the minimum_interval_seconds argument on the [@sensor](https://docs.dagster.io/_apidocs/schedules-sensors#dagster.sensor) decorator.

[Source](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#evaluation-interval)

For question 2.

In the example you provided in the URL, the job is defined in the sensor decorator.
```python
@sensor(job=my_job)
def my_s3_sensor(context):
    since_key = context.cursor or None
    new_s3_keys = get_s3_keys("my_s3_bucket", since_key=since_key)
    if not new_s3_keys:
        return SkipReason("No new s3 files found for bucket my_s3_bucket.")
    last_key = new_s3_keys[-1]
    run_requests = [RunRequest(run_key=s3_key, run_config={}) for s3_key in new_s3_keys]
    context.update_cursor(last_key)
    return run_requests
```
So in that example the job called `my_job` will be the one that get executed when you return a `run_request`.

Here's the description of the `job` parameter in the sensor decorator.
>  job = JobDefinition, UnresolvedAssetJobDefinition]]):
           The job to be executed when the sensor fires.

The `run_key` will identify the run_request in Dagster and make it unique.
The `run_config` will be push down to the job.
If you configure a `partition_key` in the `RunRequest`, then the job for that specific partition will be triggered.

**Reply by yuval-airis** [reply]
Thanks [marcilj](https://github.com/marcilj)!

Part 1 is correct, I saw it later too, thanks!

Part 2 is still unclear to me. An ID may be required, but I am confused by the fact I do not pass any parameters for the run. (e.g. In my code I have an op calculating asset C which gets assets A,B as input)

By the way, I added another question 3, but it's very specific to the S3 example

**Comment by benpankow** [comment]
Hi Yuvval - to address comments 2/3:

> Part 2 is still unclear to me. An ID may be required, but I am confused by the fact I do not pass any parameters for the run. (e.g. In my code I have an op calculating asset C which gets assets A,B as input)

The input that you may provide to a run request is run config - not the contents of any upstream assets. If you are starting a run of a job which does not run the code for upstream assets (e.g. C, not A & B), these contents would be read using [an I/O manager](https://docs.dagster.io/concepts/io-management/io-managers#using-io-managers-with-software-defined-assets).

For example, if I have assets A and B upstream of asset C, and I run a job which only materializes asset C, Dagster will assume the values of assets A and B are stored somewhere persistently (e.g. in S3, a database etc) via I/O manager, and will try to read them using the selected I/O manager's load code.

> I looked at the code of [get_s3_keys](https://github.com/dagster-io/dagster/blob/a9336275cf23ff63c0f70c1eaff5034f5bf67497/python_modules/libraries/dagster-aws/dagster_aws/s3/sensor.py#L7) and noticed it reads all keys (cursor is initially "") and only later starts updating it after every partial read. Couldn't cursor have been initialized as since_key? (Line 15)

At first glance, I agree, it looks like the current behavior is probably not the most efficient.
